src/api.js

```diff
@@ -79,48 +79,50 @@ api.declare({
     '**Warning** this api end-point is **not stable**.',
   ].join('\n'),
 }, async function(req, res) {
- 
   let {namespace} = req.params;
+  let newNamespace = await setNamespace(this, namespace);
+  
+  res.reply({
+    namespace: newNamespace.namespace,
+    username: newNamespace.username,
+    password: newNamespace.password,
+  });
+});
 
-  //check for any entries that contain the requested namespace
-  let data = await this.Namespaces.query({
-    namespace:          this.Namespaces.op.equal(namespace),
+/* Retrieve any entries containing the requested namespace.
+ * If an entry exists, use it. Otherwise, create a new entry and an associated rabbit user.*/
+async function setNamespace(context, namespace) {
+  
+  let data = await context.Namespaces.query({
+    namespace:     context.Namespaces.op.equal(namespace),
   }, {
-    limit:            250, 
-  }
-  );
-
-  var newNamespace;
-
-  if (data.entries.length === 0) {
-    //create a new entry if none exists 
-    
-    newNamespace = await this.Namespaces.create({
-      namespace: namespace,
-      username: slugid.v4(),
-      password: slugid.v4(),
-      created:  new Date(),
-      expires:  taskcluster.fromNow('1 day'),
-    });
-
-    await this.rabbit.createUser(newNamespace.username, newNamespace.password, ['taskcluster-pulse']);
+    limit:         250, 
+  });
 
-  } else if (data.entries.length === 1) { 
-    //if a namespace already exists, use the loaded username & password
+  let newNamespace;
+  let entryDNE = data.entries.length === 0;
+  
+  if (entryDNE) {
+    newNamespace = createTableEntry(context, namespace);
+    await context.rabbit.setUserPermissions(
+      user =  await context.rabbit.createUser(newNamespace.username, newNamespace.password, ['taskcluster-pulse']),
+      vhost             =  '/',
+      configurePattern  = '',
+      writePattern      = 'taskcluster/(exchanges|queues)/' + newNamespace.namespace + '/.*',
+      readPattern       = 'taskcluster/exchanges/.*'
+    ); 
+  } else { 
     newNamespace = data.entries[0]; 
-  } else {
-    throw new Error('Exacly one namespace must exist');
-  }
-
-  res.reply(
-    //return the namespace entity
-    {
-      namespace: newNamespace.namespace,
-      username: newNamespace.username,
-      password: newNamespace.password,
-    }
-   
-  );
-
-});
+  } 
+  return newNamespace;
+}
```

Comment:
So, I think this still contains a race condition.  Imagine:

 * request 1 for a new namespace N arrives
 * request 1 runs the query, gets zero results
 * request 2 for N arrives
 * request 2 runs the query, gets zero results
 * request 1 creates the table entry
 * request 2 creates the table entry ---> failure

In other cases, such as [creating a new task](https://github.com/taskcluster/taskcluster-queue/blob/master/src/api.js#L629), we try to insert into the table without checking first, and if we get a conflict error back from Azure, look at the existing record.  If it matches the parameters of the request (a trivial check for namespaces, since `namespace` is the only user-specified value in the table), then we just return that existing record; otherwise we bubble up the conflict as the two requests weren't the same.